src/mp4.zig

```diff
@@ -0,0 +1,212 @@
+const std = @import("std");
+const Allocator = std.mem.Allocator;
+const Metadata = @import("metadata.zig").Metadata;
+
+// Some atoms can be "full atoms", meaning they have an additional 4 bytes
+// for a version and some flags.
+const FullAtomHeader = struct {
+    version: u8,
+    flags: u24,
+};
+
+fn readFullAtomHeader(reader: anytype) !FullAtomHeader {
+    return FullAtomHeader{
+        .version = try reader.readByte(),
+        .flags = try reader.readIntBig(u24),
+    };
+}
+
+/// Every atom in a MP4 file has this fixed-size header
+const AtomHeader = struct {
+    /// the atom size (including the header size of 8 bytes)
+    size: u32,
+    /// the name or type
+    name: [4]u8,
+};
+
+fn readAtomHeader(reader: anytype, seekable_stream: anytype) !AtomHeader {
+    const size = switch (try reader.readIntBig(u32)) {
+        0 => blk: {
+            // a size of 0 means the atom extends to end of file
+
+            const remaining = (try seekable_stream.getEndPos()) - (try seekable_stream.getPos());
+
+            break :blk @intCast(u32, remaining);
+        },
+        1 => {
+            // a size of 1 means the atom header has an extended size field
+            // TODO: implement this if relevant ?
+
+            std.debug.panic("unhandled 64 bit size", .{});
```

Comment:
The goal is to never induce a panic regardless of the input, so this should be something like:

```suggestion
            // TODO: implement this if relevant ?

            return error.UnimplementedSizeFormat;
```